### PR TITLE
api: add TextDecoration#withState and add missing null pointer check

### LINKED ISSUE #431: Provide an alternative for `TextDecoration#as(boolean)`
On Kotlin, `as` is a protected keyword which means you can't do `TextDecoration.ITALIC.as(false)` without doing the ugly ``TextDecoration.ITALIC.`as`(false)``.

This could either exist in the API or in the Kotlin module, although preferably the former.

### PR DESCRIPTION
**Pull request by astei**
Fixes #431. While we're at it, I noticed there was no null check for the state in the `TextDecorationAndStateImpl` constructor so I fixed that too.

### CODE REVIEW

File: api/src/main/java/net/kyori/adventure/text/format/TextDecoration.java [hunk 1]
```diff
@@ -99,6 +99,28 @@
     return new TextDecorationAndStateImpl(this, state);
   }
 
+  /**
+   * An alias for {@link #as(boolean)}.
+   *
+   * @param state the state
+   * @return a {@link TextDecorationAndState}
+   * @since 4.8.0
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/text/format/TextDecoration.java:107]:
```suggestion
   * @since 4.10.0
```

File: api/src/main/java/net/kyori/adventure/text/format/TextDecoration.java [hunk 2]
```diff
@@ -99,6 +99,28 @@
...
+  public final @NotNull TextDecorationAndState withState(final boolean state) {
+    return this.as(State.byBoolean(state));
+  }
+
+  /**
+   * An alias for {@link #as(State)}.
+   *
+   * @param state the state
+   * @return a {@link TextDecorationAndState}
+   * @since 4.8.0
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/text/format/TextDecoration.java:118]:
```suggestion
   * @since 4.10.0
```

**Review by kashike — CHANGES_REQUESTED** [review]

**Review by kezz — APPROVED** [review]
I wonder if we should deprecate the `as` method because `withState` is much clearer? Either way I'm happy though.

File: api/src/main/java/net/kyori/adventure/text/format/TextDecoration.java [hunk 3]
```diff
@@ -99,6 +99,28 @@
...
+  /**
+   * An alias for {@link #as(State)}.
+   *
+   * @param state the state
+   * @return a {@link TextDecorationAndState}
+   * @since 4.10.0
+   */
+  public final @NotNull TextDecorationAndState withState(final @NotNull State state) {
+    return this.as(state);
+  }
```
Comment (zml2008) [inline api/src/main/java/net/kyori/adventure/text/format/TextDecoration.java:122]: would you be able to add `withState(TriState)` (and `as(TriState)`) overloads here? kinda scope creep but merging #495 sparked the thought

**Review by zml2008 — CHANGES_REQUESTED** [review]

File: api/src/main/java/net/kyori/adventure/text/format/TextDecoration.java (hunk 3 above)
Comment (kezz) [reply api/src/main/java/net/kyori/adventure/text/format/TextDecoration.java:122]: Do we need an `as(TriState)` overload? As it's new API we don't need another method here that does the same thing. Although I'm also for deprecating the existing `as` method, so that might change if we decide not to.

**Review by zml2008 — APPROVED** [review]

**Review by kashike — APPROVED** [review]
